org.jacoco.examples/src/org/jacoco/examples/XMLDiff.java

```diff
@@ -0,0 +1,231 @@
+									+ sb.toString());
+						}
+					}
+				});
+			}
+		});
+	}
+
+	private interface Callback<K, V> {
+
+		void inLeft(K key, V value);
+
+		void inRight(K key, V value);
+
+		void inBoth(K key, V left, V right);
+
+	}
+
+	private static <K, V> void compare(final Map<K, V> left,
+			final Map<K, V> right, final Callback<K, V> callback) {
+		final Set<K> keys = new HashSet<K>();
+		keys.addAll(left.keySet());
+		keys.addAll(right.keySet());
+		for (K key : keys) {
+			final V leftValue = left.get(key);
+			final V rightValue = right.get(key);
+			if (leftValue == null) {
+				callback.inRight(key, rightValue);
+			} else if (rightValue == null) {
+				callback.inLeft(key, leftValue);
+			} else {
+				callback.inBoth(key, leftValue, rightValue);
+			}
+		}
+	}
+
+	/**
+	 * Reads JaCoCo XML report.
+	 *
+	 * @param in
+	 *            stream to read XML report from
+	 * @return parsed report
+	 * @throws Exception
+	 *             in case of errors
+	 */
+	static Map<String, Map<Integer, LineInfo>> read(final InputStream in)
+			throws Exception {
+		final SAXParserFactory factory = SAXParserFactory.newInstance();
+		final SAXParser parser = factory.newSAXParser();
+		final ReportHandler handler = new ReportHandler();
+		parser.parse(in, handler);
+		return handler.r;
+	}
+
+	static final class LineInfo {
+		final int mi;
+		final int ci;
+		final int mb;
+		final int cb;
+
+		LineInfo(final int mi, final int ci, final int mb, final int cb) {
+			this.mi = mi;
+			this.ci = ci;
+			this.mb = mb;
+			this.cb = cb;
+		}
+	}
+
+	private static class ReportHandler extends DefaultHandler {
+		final Map<String, Map<Integer, LineInfo>> r = new HashMap<String, Map<Integer, LineInfo>>();
+
+		private String packageName;
+		private Map<Integer, LineInfo> s;
+
+		@Override
+		public void startElement(final String uri, final String localName,
+				final String qName, final Attributes attributes) {
+			if ("package".equals(qName)) {
+				packageName = attributes.getValue("name");
+
+			} else if ("sourcefile".equals(qName)) {
+				final String fileName = attributes.getValue("name");
+				final String name = packageName + "/" + fileName;
+				s = new HashMap<Integer, LineInfo>();
+				if (r.put(name, Collections.unmodifiableMap(s)) != null) {
+					throw new IllegalStateException("duplicate " + name);
+				}
+
+			} else if ("line".equals(qName)) {
+				final int lineNumber = Integer
+						.parseInt(attributes.getValue("nr"));
+				final int mi = Integer.parseInt(attributes.getValue("mi"));
+				final int ci = Integer.parseInt(attributes.getValue("ci"));
+				final int mb = Integer.parseInt(attributes.getValue("mb"));
+				final int cb = Integer.parseInt(attributes.getValue("cb"));
+				s.put(lineNumber, new LineInfo(mi, ci, mb, cb));
+			}
+		}
+
+		@Override
+		public InputSource resolveEntity(final String publicId,
+				final String systemId) {
+			if (publicId.startsWith("-//JACOCO//DTD Report")) {
+				return new InputSource(new StringReader(""));
```

Comment:
This is an example and should follow best practices:

- check for exact DTD version to ensure compatibility
- Supply actual DTD otherwise this will fail when we merge #813

Reply:
Sorry, my bad: I overlooked that `parseAsOptionalInt()` already checks for `null` values.